index_update.go

```diff
@@ -0,0 +1,625 @@
+		return nil, err
+	}
+
+	oriPaths := make(map[string]*pathInfo)
+	updPaths := make(map[string]*pathInfo)
+
+	// Go through each mapping present in the original
+	// and consolidate according to the document paths
+	for name, oriDMapping := range ori.TypeMapping {
+		addPathInfo(oriPaths, "", oriDMapping, ori, nil, name)
+	}
+	addPathInfo(oriPaths, "", ori.DefaultMapping, ori, nil, "")
+
+	// Go through each mapping present in the updated
+	// and consolidate according to the document paths
+	for name, updDMapping := range upd.TypeMapping {
+		addPathInfo(updPaths, "", updDMapping, ori, nil, name)
+	}
+	addPathInfo(updPaths, "", upd.DefaultMapping, ori, nil, "")
+
+	// Compare all components of custom analysis currently in use
+	err = compareCustomComponents(oriPaths, updPaths, ori, upd)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compare both the mappings based on the document paths
+	// and create a list of index, docvalues, store differences
+	// for every single field possible
+	fieldInfo := make(map[string]*index.UpdateFieldInfo)
+	for path, info := range oriPaths {
+		err = addFieldInfo(fieldInfo, info, updPaths[path])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Remove entries from the list with no changes between the
+	// original and the updated mapping
+	for name, info := range fieldInfo {
+		if !info.Deleted && !info.Index && !info.DocValues && !info.Store {
+			delete(fieldInfo, name)
+		}
+		// A field cannot be completely deleted with any dynamic value turned on
+		if info.Deleted {
+			if upd.IndexDynamic {
+				return nil, fmt.Errorf("Mapping cannot be removed when index dynamic is true")
+			}
+			if upd.StoreDynamic {
+				return nil, fmt.Errorf("Mapping cannot be removed when store dynamic is true")
+			}
+			if upd.DocValuesDynamic {
+				return nil, fmt.Errorf("Mapping cannot be removed when docvalues dynamic is true")
+			}
+		}
+	}
+	return fieldInfo, nil
+}
+
+// Ensures non of the top level index mapping fields have changed
+func compareMappings(ori, upd *mapping.IndexMappingImpl) error {
+	if ori.TypeField != upd.TypeField &&
+		(len(ori.TypeMapping) != 0 || len(upd.TypeMapping) != 0) {
+		return fmt.Errorf("type field cannot be changed when type mappings are present")
+	}
+
+	if ori.DefaultType != upd.DefaultType {
+		return fmt.Errorf("default type cannot be changed")
+	}
+
+	if ori.DefaultField != upd.DefaultField {
+		return fmt.Errorf("default field cannot be changed")
+	}
+
+	if ori.IndexDynamic != upd.IndexDynamic {
+		return fmt.Errorf("index dynamic cannot be changed")
+	}
+
+	if ori.StoreDynamic != upd.StoreDynamic {
+		return fmt.Errorf(("store dynamic cannot be changed"))
+	}
+
+	if ori.DocValuesDynamic != upd.DocValuesDynamic {
+		return fmt.Errorf(("docvalues dynamic cannot be changed"))
+	}
+
+	return nil
+}
+
+// Ensures updated document mapping does not contain new
+// field mappings or document mappings
+func checkUpdatedMapping(ori, upd *mapping.DocumentMapping) error {
+
```

Comment:
small nit: could you please remove these new lines (i see a bunch of them) just so that we are consistent with the other functions' structure